### PR TITLE
Support for Annotated Timeseries

### PR DESCRIPTION
**Pull request by williamsdoug**
Extends TSTensor, TSDataset(s) and TSDataLoader(s) by supporting annotated points of interest.

Features:
- Extracts subsequence of X value in proximity of annotation point, as specified by window and offset
- Flexible support of y values
  - Common-per-row labels
  - Per-annotation labels
  - If Y is numeric vector or array, extracts subsequence using same window parameters as X
- Static and dynamic modes of operation
  - Static:  Preprocessing function, includes support for mapping folds and splits
  - Dynamic: TSAnnotatedTensor,  TSAnnotatedDataset(s), TSAnnotatedDataLoader(s), supporting transforms and convenience functions
-   Limitations:  TSAnnotatedDataLoaders does not currently support adding Test dataset

### CODE REVIEW

**Comment by oguiza** [comment]
Hi @williamsdoug,

Thanks for putting together this comprehensive, high-quality PR. The level of understanding you have of `tsai` is really amazing!

 I can see the value of your approach, but there are multiple downsides to it. However, I don't think that this type of functionality should be integrated into `tsai` now for the reasons expressed in issue #85 [(comment)](https://github.com/timeseriesAI/tsai/issues/85#issuecomment-807163837).

I'll expand my concerns here: 

- As I stated in that message, data preparation is outside the scope of `tsai`.
- I have included some functionality to demonstrate how data can be prepared to be used in `tsai` that is very commonly used (sliding window type of approaches, multiple files, etc).
- What you are proposing seems a very customized approach. Somebody else asked for some similar functionality, but you couldn't get alignment on a single design, as your needs were different. 
- The proposed solution requires changes in almost every aspect of `tsai`, from the tensor type, dataset, dataloader, etc. Based on my experience this type of change is then difficult to maintain. When I need to make a change in `tsai` this will likely have a knock-on effect on this functionality. 
- I'm not sure that when another user needs this type of data preparation, he will able to use this as-is. 

Based on all this I'd propose the following. You already have the code you need in your repo. You can start using it. It's likely that when using it you'll find areas that need to be tweaked. You can make changes to it as required (it's your own code). In the meantime, let's see if there's a higher demand for this type of solution. If there's more demand, and if the proposed solution is general enough to be able to meet those needs, we can then add the code to `tsai`.

**Comment by williamsdoug** [comment]
Hi @oguiza.  Thanks for the feedback, and I share your concern about number of classes touched and it's impact on maintainability.

Just as you've recently extended tsai to include dynamic (training-time) sliding window, I believe that there is value in tsai including the concept of user-specified regions of interest, where the region of interest may also include an associated label(s).  Much of the complexity of the current implementation is related to the inclusion of annotation metadata with the X value (TSAnnotatedTensor), resulting in the various new Tensor, Dataset and Dataloader classes.  If instead the annotation metadata was included along with the y values (ie: annotated labels), I believe this feature could be implemented primarily as a Transform similar to  TSAnnotationWindow included with my prototype implementation.

My concern over generalizing to include the Session Window use case (issue #85) is that much of the required data manipulation for session windows seems to occur within the Pandas dataframe domain.  Once the problem is transformed into something that can be expressed as a Numpy array or PyTorch Tensor,  there may be commonality with the solutions I've proposed (extraction of user-defined windows with labels).

Let me know if this might be of interest.  I'f so, I'll pursue further in a more general form.

**Comment by williamsdoug** [comment]
Moving  content to repo williamsdoug/tsai_ecg after refactoring to reduce number of subclassed  tsai datatypes.  Feel free to pull into tsai if any of the extensions are more generally useful
